substrate/frame/multi-asset-bounties/src/lib.rs

```diff
@@ -0,0 +1,1962 @@
+				if let Some(curator_deposit) = maybe_curator_deposit {
+					bounty.curator_deposit = curator_deposit;
+				}
+				if let Some(fee) = maybe_fee {
+					bounty.fee = fee;
+				}
+				Bounties::<T, I>::insert(parent_bounty_id, bounty);
+				Ok(())
+			},
+			Some(child_bounty_id) => {
+				let mut bounty = ChildBounties::<T, I>::get(parent_bounty_id, child_bounty_id)
+					.ok_or(Error::<T, I>::InvalidIndex)?;
+				bounty.status = new_status;
+				if let Some(curator_deposit) = maybe_curator_deposit {
+					bounty.curator_deposit = curator_deposit;
+				}
+				if let Some(fee) = maybe_fee {
+					bounty.fee = fee;
+				}
+				ChildBounties::<T, I>::insert(parent_bounty_id, child_bounty_id, bounty);
+				Ok(())
+			},
+		}
+	}
+
+	/// Calculates amount the beneficiary and curator stash during child-/bounty payout.
+	fn calculate_curator_fee_and_payout(
+		parent_bounty_id: BountyIndex,
+		fee: BalanceOf<T, I>,
+		value: BalanceOf<T, I>,
+	) -> (BalanceOf<T, I>, BalanceOf<T, I>) {
+		// Get total child bounties curator fees, and subtract it from the parent
+		// curator fee (the fee in present referenced bounty, `self`).
+		let children_fee = ChildBountiesCuratorFeesPerParent::<T, I>::get(parent_bounty_id);
+		debug_assert!(children_fee <= fee);
+		let final_fee = fee.saturating_sub(children_fee);
+
+		// Get total child bounties value, and subtract it from the parent
+		// value (the value in present referenced bounty, `self`).
+		let children_value = ChildBountiesValuePerParent::<T, I>::get(parent_bounty_id);
+		debug_assert!(children_value <= value);
+		let value_remaining = value.saturating_sub(children_value);
+		let payout = value_remaining.saturating_sub(final_fee);
+
+		(final_fee, payout)
+	}
+
+	/// Cleanup a child-/bounty from the storage.
+	fn remove_bounty(parent_bounty_id: BountyIndex, child_bounty_id: Option<BountyIndex>) {
+		match child_bounty_id {
+			None => {
+				Bounties::<T, I>::remove(parent_bounty_id);
+				BountyDescriptions::<T, I>::remove(parent_bounty_id);
+				ChildBountiesPerParent::<T, I>::remove(parent_bounty_id);
+				TotalChildBountiesPerParent::<T, I>::remove(parent_bounty_id);
+			},
+			Some(child_bounty_id) => {
+				ChildBounties::<T, I>::remove(parent_bounty_id, child_bounty_id);
+				ChildBountyDescriptions::<T, I>::remove(parent_bounty_id, child_bounty_id);
+				ChildBountiesPerParent::<T, I>::mutate(parent_bounty_id, |count| {
+					count.saturating_dec()
+				});
+			},
+		}
+	}
+
+	/// Initiates payment from the funding source to the child-/bounty account/location.
+	fn do_process_funding_payment(
+		parent_bounty_id: BountyIndex,
+		child_bounty_id: Option<BountyIndex>,
+		asset_kind: T::AssetKind,
+		value: BalanceOf<T, I>,
+		maybe_payment_status: Option<PaymentState<PaymentIdOf<T, I>>>,
+	) -> Result<PaymentState<PaymentIdOf<T, I>>, DispatchError> {
+		if let Some(payment_status) = maybe_payment_status {
+			ensure!(payment_status.is_pending_or_failed(), Error::<T, I>::UnexpectedStatus);
```

Comment:
I would expect to ensure it is `Failed`, if the status is pending we don't want to do another payment no?

this function is called both on the funding calls and `retry_payment`. but retry payment should only by retried after a failure I guess.